airflow/contrib/hooks/gcp_vision_hook.py

```diff
@@ -0,0 +1,245 @@
+    def update_product_set(self, product_set, location=None, product_set_id=None, update_mask=None,
+                           project_id=None, retry=None, timeout=None, metadata=None):
+        """
+        For the documentation see:
+        :py:class:`~airflow.contrib.operators.gcp_vision_operator.CloudVisionProductSetUpdateOperator`
+        """
+        client = self.get_client()
+        product_set = self._determine_name(product_set, product_set_id, location, project_id)
+        self.log.info('Updating ProductSet: %s', product_set.name)
+        response = self._handle_request(lambda **kwargs: client.update_product_set(**kwargs),
+                                        product_set=product_set, update_mask=update_mask, retry=retry,
+                                        timeout=timeout, metadata=metadata)
+        self.log.info('ProductSet updated: %s', response.name if response else '')
+        self.log.debug('ProductSet updated:\n%s', response)
+        return MessageToJson(response)
+
+    @GoogleCloudBaseHook.fallback_to_default_project_id
+    def delete_product_set(self, location, product_set_id, project_id=None, retry=None, timeout=None,
+                           metadata=None):
+        """
+        For the documentation see:
+        :py:class:`~airflow.contrib.operators.gcp_vision_operator.CloudVisionProductSetDeleteOperator`
+        """
+        client = self.get_client()
+        name = PRODUCT_SET_NAME_TEMPLATE.format(project_id, location, product_set_id)
+        self.log.info('Deleting ProductSet: %s', name)
+        response = self._handle_request(lambda **kwargs: client.delete_product_set(**kwargs), name=name,
+                                        retry=retry, timeout=timeout, metadata=metadata)
+        self.log.info('ProductSet with the name [%s] deleted.', name)
+        return response
+
+    @staticmethod
+    def _determine_name(entity, entity_id, location, project_id):
+        is_product = type(entity) == Product
+        entity_label = 'Product' if is_product else 'ProductSet'
+        entity_id_label = 'product_id' if is_product else 'productset_id'
+        entity_name_template = PRODUCT_NAME_TEMPLATE if is_product else PRODUCT_SET_NAME_TEMPLATE
+        explicit_name = getattr(entity, 'name')
+        if not location or not entity_id:
+            # Not enough parameters to construct the name. Trying to use the name from Product.
+            if not explicit_name:
+                raise AirflowException("Unable to determine the {} name. Please either set the name "
+                                       "directly in the {} object or provide the `location` and "
+                                       "`{}` parameters.".format(entity_label, entity_label, entity_id_label))
+        else:
+            template_name = entity_name_template.format(project_id, location, entity_id)
+            if not explicit_name:
+                entity.name = template_name
```

Comment:
I think, you should avoid modify objects from user.  He may not know about it and want to use the same object in the loop.
https://softwareengineering.stackexchange.com/questions/245767/is-modifying-an-incoming-parameter-an-antipattern

Reply:
I can agree on that. I effectively mixed two things: I do return the full object (see https://github.com/PolideaInternal/airflow/pull/51/files/e6868c2bb5d0154891dd0f8b0f1b08292c648b98#r257625543), which is a good approach, but I also changed the object passed in by the user.

So I added a deepcopy at the beginning of this function, and this way I operate on a copy of the object, thus not modifying the user's object, and at the end of the function I return the modified object to the user.

`entity = deepcopy(entity)`